packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1970 @@
+    const methodImplementation = this._getRestrictedMethod(methodName, origin);
+
+    const result = await this._executeRestrictedMethod(
+      methodImplementation,
+      { origin },
+      methodName,
+      params,
+    );
+
+    if (result === undefined) {
+      throw new Error(
+        `Internal request for method "${methodName}" as origin "${origin}" returned no result.`,
+      );
+    }
+
+    return result;
+  }
+
+  /**
+   * An internal method used in the controller's `json-rpc-engine` middleware
+   * and {@link PermissionController.executeRestrictedMethod}. Calls the
+   * specified restricted method implementation after decorating it with the
+   * caveats of its permission. Throws if the subject does not have the
+   * requisite permission.
+   *
+   * **ATTN:** Assumes that all parameters are valid.
+   *
+   * @see {@link PermissionController.executeRestrictedMethod} and
+   * {@link PermissionController.createPermissionMiddleware} for usage.
+   * @param methodImplementation - The implementation of the method to call.
+   * @param subject - Metadata about the subject that made the request.
+   * @param req - The request object associated with the request.
+   * @returns
+   */
+  private _executeRestrictedMethod(
+    methodImplementation: RestrictedMethod<RestrictedMethodParameters, Json>,
+    subject: PermissionSubjectMetadata,
+    method: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    params: RestrictedMethodParameters = [],
+  ): ReturnType<RestrictedMethod<RestrictedMethodParameters, Json>> {
+    const { origin } = subject;
+
+    const permission = this.getPermission(origin, method);
+    if (!permission) {
+      throw unauthorized({ data: { origin, method } });
+    }
+
+    return decorateWithCaveats<ControllerCaveatSpecification>(
+      methodImplementation,
+      permission,
+      this._caveatSpecifications,
+    )({ method, params, context: { origin } });
+  }
+
+  /**
+   * Like {@link PermissionController.getRestrictedMethod}, except the request
+   * for the restricted method is associated with a specific subject, and a
+   * JSON-RPC error is thrown if the method does not exist.
+   *
+   * @see {@link PermissionController.executeRestrictedMethod} and
+   * {@link PermissionController.createPermissionMiddleware} for usage.
+   * @param method - The name of the method to get.
+   * @param request - The request associated with the request for the method.
+   * @returns - The restricted method implementation, if it exists.
+   */
+  private _getRestrictedMethod(
+    method: string,
+    origin: OriginString,
+  ): RestrictedMethod<RestrictedMethodParameters, Json> {
+    const methodImplementation = this.getRestrictedMethod(
+      method as ExtractPermission<
+        ControllerPermissionSpecification,
+        ControllerCaveatSpecification
+      >['parentCapability'],
+    );
```

Comment:
Shouldn't the type be applied to the parameter, rather than asserted using a cast? I might be missing something here but this seems like a strange pattern.

```suggestion
    method: ExtractPermission<
      ControllerPermissionSpecification,
      ControllerCaveatSpecification
    >['parentCapability'],
    origin: OriginString,
  ): RestrictedMethod<RestrictedMethodParameters, Json> {
    const methodImplementation = this.getRestrictedMethod(method);
```

Reply:
Ah, yes, that is weird. The cast turned out to be pointless because `getRestrictedMethod` already accepted any string. It contained the same cast to call `getTargetKey`, but all of these methods are used to check if a permission name refers to a specified permission, so the solution is to make all of them accept strings.

e1e50dd